src/activity-logger/background/index.js

```diff
@@ -18,23 +21,57 @@ function logPageVisit({url}) {
     return storeVisit({timestamp, visitInfo, pageInfo})
 }
 
-// Listen for navigations to log them and analyse the pages.
-browser.webNavigation.onCommitted.addListener(details => {
+let pageLoaded = function(item){
+    pages.forEach((page)=>{
+        if(item === page){
+            return true;
+        }
+    });
+    return false;
+}
+
+let navigated = function(details){
+     console.log(details);
     // Ignore pages loaded in frames, it is usually noise.
     if (details.frameId !== 0)
         return
 
     if (!isWorthRemembering({url: details.url}))
         return
 
+    // check to make sure the page has not been loaded before by any of the listeners.To avoid double loading
+    if(pageLoaded({tabId: details.tabId, url: details.url})){
+        return
+    }
+
     // Consider every navigation a new visit.
     logPageVisit({
         url: details.url
-    }).then(
+    }).then(        
         // Wait until its DOM has loaded.
         value => whenPageDOMLoaded({tabId: details.tabId}).then(()=>value)
     ).then(({visitId, pageId}) => {
+
+        //add the loaded page to the array to prevent double loading due to double checks
+        pages.push({tabId: details.tabId, url: details.tabId});
+
         // Start page content analysis (text extraction, etcetera)
         performPageAnalysis({pageId, tabId: details.tabId})
-    })
-})
+    });
+}
+
+
+
+
+// Listen for navigations to log them and analyse the pages.
+browser.webNavigation.onCommitted.addListener((details)=>navigated(details))
+
+// Listen for navigations in web apps to log them and analyse the pages.
+browser.webNavigation.onHistoryStateUpdated.addListener((details)=>{
+    browser.tabs.get(details.tabId,(tab)=>{
+        
+        //make sure the page logged is the nal page that is reached by the history change action
+        if(tab.url===details.url)
+            navigated(details)
```

Comment:
I noticed in some cases, like youtube, when a video was clicked, the onHistoryStateChanges was fired twice, so i needed to check to make sure that only the right page was taken.